syntax = "proto3";


message Sensor {
    message Owner {
        // Public key of the user who owns the sensor
        string user_public_key = 1;

        // Approximately when the owner was updated, as a Unix UTC timestamp
        uint64 timestamp = 2;
    }

    message Location {
        // Coordinates are expected to be in millionths of a degree
        sint64 latitude = 1;
        sint64 longitude = 2;

        // Approximately when the location was updated, as a Unix UTC timestamp
        uint64 timestamp = 3;
    }

    message Measurement {
        // Measurement value
        double measurement = 1;

        // Approximately when the measurement took place, such as
        // Unix UTC timestamp
        uint64 timestamp = 2;
    }

    // The user-defined natural key which identifies the object in the
    // real world (for example a serial number)
    string sensor_id = 1;

    // Ordered oldest to newest by timestamp
    repeated Owner owners = 2;
    repeated Location locations = 3;
    repeated Measurement measurements = 4;
}


message SensorContainer {
    repeated Sensor entries = 1;
}